File: src/main/java/in/dragonbra/javasteam/steam/handlers/steamapps/callback/FreeLicenseCallback.java
package in.dragonbra.javasteam.steam.handlers.steamapps.callback;

import in.dragonbra.javasteam.enums.EResult;
import in.dragonbra.javasteam.protobufs.steamclient.SteammessagesClientserver2.CMsgClientRequestFreeLicenseResponse;
import in.dragonbra.javasteam.steam.handlers.steamapps.SteamApps;
import in.dragonbra.javasteam.steam.steamclient.callbackmgr.CallbackMsg;
import in.dragonbra.javasteam.types.JobID;

import java.util.Collections;
import java.util.List;

/**
 * This callback is received in response to calling {@link SteamApps#requestFreeLicense}, informing the client of newly granted packages, if any.
 */
public class FreeLicenseCallback extends CallbackMsg {

    private EResult result;

    private List<Integer> grantedApps;

    private List<Integer> grantedPackages;

    public FreeLicenseCallback(JobID jobID, CMsgClientRequestFreeLicenseResponse.Builder msg) {
        setJobID(jobID);

        result = EResult.from(msg.getEresult());

        grantedApps = Collections.unmodifiableList(msg.getGrantedAppidsList());
        grantedPackages = Collections.unmodifiableList(msg.getGrantedPackageidsList());
    }

    public EResult getResult() {
        return result;
    }

    public List<Integer> getGrantedApps() {
        return grantedApps;
    }

    public List<Integer> getGrantedPackages() {
        return grantedPackages;
    }
}

File: src/main/java/in/dragonbra/javasteam/steam/handlers/steamapps/callback/FreeLicenseCallback.kt
package `in`.dragonbra.javasteam.steam.handlers.steamapps.callback

import `in`.dragonbra.javasteam.base.ClientMsgProtobuf
import `in`.dragonbra.javasteam.base.IPacketMsg
import `in`.dragonbra.javasteam.enums.EResult
import `in`.dragonbra.javasteam.protobufs.steamclient.SteammessagesClientserver2.CMsgClientRequestFreeLicenseResponse
import `in`.dragonbra.javasteam.steam.handlers.steamapps.SteamApps
import `in`.dragonbra.javasteam.steam.steamclient.callbackmgr.CallbackMsg
import java.util.*

/**
 * This callback is received in response to calling [SteamApps.requestFreeLicense], informing the client of newly granted packages, if any.
 */
@Suppress("MemberVisibilityCanBePrivate")
class FreeLicenseCallback(packetMsg: IPacketMsg) : CallbackMsg() {

    /**
     * Gets the result of the message.
     */
    val result: EResult

    /**
     * Gets the list of granted apps.
     */
    val grantedApps: List<Int>

    /**
     * Gets the list of granted packages.
     */
    val grantedPackages: List<Int>

    init {
        val grantedLicenses = ClientMsgProtobuf<CMsgClientRequestFreeLicenseResponse.Builder>(
            CMsgClientRequestFreeLicenseResponse::class.java,
            packetMsg
        )
        val msg = grantedLicenses.body

        jobID = grantedLicenses.targetJobID

        result = EResult.from(msg.eresult)

        grantedApps = Collections.unmodifiableList(msg.grantedAppidsList)
        grantedPackages = Collections.unmodifiableList(msg.grantedPackageidsList)
    }
}
